src/backend/distributed/utils/reference_table_utils.c

```diff
@@ -0,0 +1,247 @@
+ * metadata to mark the table as reference.
+ */
+Datum
+upgrade_to_reference_table(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	List *shardIntervalList = NIL;
+	ShardInterval *shardInterval = NULL;
+	uint64 shardId = INVALID_SHARD_ID;
+
+	if (!IsDistributedTable(relationId))
+	{
+		char *relationName = get_rel_name(relationId);
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						errmsg("cannot upgrade to reference table"),
+						errdetail("Relation \"%s\" is not distributed.", relationName),
+						errhint("Instead, you can use; "
+								"create_reference_table('%s');", relationName)));
+	}
+
+	if (PartitionMethod(relationId) == DISTRIBUTE_BY_NONE)
+	{
+		char *relationName = get_rel_name(relationId);
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						errmsg("cannot upgrade to reference table"),
+						errdetail("Relation \"%s\" is already a reference table",
+								  relationName)));
+	}
+
+	shardIntervalList = LoadShardIntervalList(relationId);
+	if (list_length(shardIntervalList) != 1)
+	{
+		char *relationName = get_rel_name(relationId);
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot upgrade to reference table"),
+						errdetail("Relation \"%s\" shard count is not one. Only "
+								  "relations with one shard can be upgraded to "
+								  "reference tables.", relationName)));
+	}
+
+	shardInterval = (ShardInterval *) linitial(shardIntervalList);
+	shardId = shardInterval->shardId;
+
+	LockShardDistributionMetadata(shardId, ExclusiveLock);
+	LockShardResource(shardId, ExclusiveLock);
+
+	ReplicateSingleShardTableToAllWorkers(relationId);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * ReplicateSingleShardTableToAllWorkers accepts a broadcast table and replicates it to
+ * all worker nodes. It assumes that caller of this function ensures that given broadcast
+ * table has only one shard.
+ */
+static void
+ReplicateSingleShardTableToAllWorkers(Oid relationId)
+{
+	List *shardIntervalList = LoadShardIntervalList(relationId);
+	ShardInterval *shardInterval = (ShardInterval *) linitial(shardIntervalList);
+	uint64 shardId = shardInterval->shardId;
+
+	List *foreignConstraintCommandList = CopyShardForeignConstraintCommandList(
+		shardInterval);
+
+	if (foreignConstraintCommandList != NIL || TableReferenced(relationId))
+	{
+		char *relationName = get_rel_name(relationId);
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot upgrade to reference table"),
+						errdetail("Relation \"%s\" is part of a foreign constraint. "
+								  "Foreign key constraints are not allowed "
+								  "from or to reference tables.", relationName)));
+	}
+
+	/*
+	 * ReplicateShardToAllWorkers function opens a transaction at each worker and
+	 * replicates given shard to all workers. If a worker already has a healthy replica
+	 * of given shart, it skips that worker to prevent copying unnecessary data.
```

Comment:
`a transaction at each worker ` -> `separate transactions (i.e., not part of any coordinated transactions) to each worker`